library/psa_crypto.c

```diff
@@ -101,26 +102,60 @@ typedef struct {
 
 static psa_global_data_t global_data;
 
+static uint8_t psa_get_initialized(void)
+{
+    uint8_t initialized;
+
+#if defined(MBEDTLS_THREADING_C)
+    mbedtls_mutex_lock(&mbedtls_threading_psa_globaldata_mutex);
+#endif /* defined(MBEDTLS_THREADING_C) */
+
+    initialized = global_data.initialized;
+
+#if defined(MBEDTLS_THREADING_C)
+    mbedtls_mutex_unlock(&mbedtls_threading_psa_globaldata_mutex);
+#endif /* defined(MBEDTLS_THREADING_C) */
+
+    return initialized;
+}
+
+static uint8_t psa_get_drivers_initialized(void)
+{
+    uint8_t initialized;
+
+#if defined(MBEDTLS_THREADING_C)
+    mbedtls_mutex_lock(&mbedtls_threading_psa_globaldata_mutex);
+#endif /* defined(MBEDTLS_THREADING_C) */
+
+    initialized = global_data.drivers_initialized;
+
+#if defined(MBEDTLS_THREADING_C)
+    mbedtls_mutex_unlock(&mbedtls_threading_psa_globaldata_mutex);
+#endif /* defined(MBEDTLS_THREADING_C) */
+
+    return initialized;
+}
+
 #if !defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
 mbedtls_psa_drbg_context_t *const mbedtls_psa_random_state =
     &global_data.rng.drbg;
```

Comment:
Unfortunately, given the way that the state is passed around as a pointer, I don't see how I could protect all the uses. There is already a mutex within `mbedtls_ctr_drbg_context` which should protect a lot of these uses, but otherwise its going to be extremely difficult to add more protection.

Reply:
When the PSA RNG is implemented on top of CTR_DRBG and HMAC_DRBG, i.e. when `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` is disabled, the PSA part of the code doesn't need to take any precautions, because the CTR_DRBG and HMAC_DRBG modules have their own mutex guarding the RNG state.

The external RNG case is [its own task](https://github.com/Mbed-TLS/mbedtls/issues/3391), which I think isn't a 3.6.0 objective?

Reply:
No, its not part of this work, thanks for clarifying that.